framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,656 @@
+        for (Annotation tla : annos) {
+            if (!isInternalJDKAnnotation(tla.def.name)) {
+                sb.append(formatAnnotation(tla));
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Formats the component types of an array via recursive descent through the array's scene-lib
+     * structure.
+     *
+     * @param e the array's scenelib type element
+     * @param arrayType the string representation of the array's type
+     * @return the type formatted to be written to Java source code, followed by a space character
+     */
+    private static String formatArrayType(ATypeElement e, String arrayType) {
+        StringBuilder result = new StringBuilder();
+        // If there are not annotations on the component, this will be true - unless there
+        // are also no annotations on the first array level. See comment below for how that
+        // case is handled.
+        int componentEndPos;
+        {
+            componentEndPos = arrayType.indexOf(' ');
+            if (arrayType.startsWith("@")) {
+                // While there are more explicit annotations, go to the next space.
+                while (arrayType.charAt(componentEndPos + 1) == '@') {
+                    componentEndPos = arrayType.indexOf(' ', componentEndPos + 1);
+                }
+                // Once all annotations have been skipped, we still have to skip to the end of the
+                // component type itself.
+                componentEndPos = arrayType.indexOf(' ', componentEndPos + 1);
+            }
+            // If the first array level doesn't have an annotation, then the spacing will be off,
+            // so use the position of the first '[' instead (which must be correct, since the
+            // first array doesn't have an annotation).
+            int firstArrayPos = arrayType.indexOf('[');
+            if (firstArrayPos < componentEndPos || componentEndPos == -1) {
+                componentEndPos = firstArrayPos - 1;
+            }
+        }
+
+        String componentType = arrayType.substring(0, componentEndPos + 1);
+        String arrayTypes = arrayType.substring(componentEndPos + 1);
+
+        return formatArrayTypeImpl(e, arrayTypes, componentType, result) + " ";
+    }
+
+    /**
+     * The implementation of formatArrayType. Java array types have a somewhat unintuitive syntax:
+     * see <a
+     * href="https://checkerframework.org/jsr308/specification/java-annotation-design.html#array-syntax">this
+     * explanation</a>.
+     *
+     * <p>The iteration order in the scene-lib representation is from outermost to innermost. For
+     * example, given the type {@code @Foo int @Bar [] @Baz []}, the iteration order is {@code @Bar
+     * []}, then {@code @Baz []}, and then finally {@code @Foo int}. This implementation therefore
+     * passes a string builder with the result of the array types seen so far, to handle
```

Comment:
This comment assumes an understanding of a set of method calls that occur over time.  Can you describe it for a single call?  I think it's the containing arrays (or array levels if you prefer that terminology).

This occurs elsewhere in the documentation, too:  some of the comments seem to be with respect to the very first call in a sequence of recursive call, and others are with respect to the current call.  Try to write it with respect to the current call.  Another way of saying that is that you should write a "loop invariant" for the loop that the recursive calls represent.  (I'm not asking you to change this to a loop; the recursive formulation is fine.)